File: ruby-gem/lib/models/physical_place.rb
# frozen_string_literal: true

# Class from DFC Business Ontology: #PhysicalPlace
require_relative '../core/semantic_object'
require_relative 'place'


module DfcLinkmlConnector
  module Models
    class PhysicalPlace < Place
      SEMANTIC_TYPE = "dfc-b:PhysicalPlace".freeze

      # @return [String]
      attr_accessor :has_phone_number

      # @return [Address, String]
      attr_accessor :address

      # @return [Feature, String]
      attr_accessor :geo_json_feature

      # @return [Person, String]
      attr_accessor :main_contact

      # @return [OpeningHoursSpecification, String]
      attr_accessor :is_open_during

      # @return [Array<TheoriticalStock, String>]
      attr_accessor :localizes

      # @return [Array<RealStock, String>]
      attr_accessor :stores

      # @param semanticId [String]
      # @param date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil, hosts: [], hasPhoneNumber: nil, address: nil, geoJsonFeature: nil, mainContact: nil, isOpenDuring: nil, localizes: [], stores: []
      def initialize(semanticId, date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil, hosts: [], hasPhoneNumber: nil, address: nil, geoJsonFeature: nil, mainContact: nil, isOpenDuring: nil, localizes: [], stores: [])
        super(semanticId, date: date, description: description, name: name, characteristicOf: characteristicOf, dimension: dimension, hosts: hosts)
        @has_phone_number = hasPhoneNumber
        @address = address
        @geo_json_feature = geoJsonFeature
        @main_contact = mainContact
        @is_open_during = isOpenDuring
        @localizes = localizes
        @stores = stores
        self.semanticType = "dfc-b:PhysicalPlace"
        registerSemanticProperty("dfc-b:hasPhoneNumber", &method("has_phone_number")).valueSetter = method("has_phone_number=")
        registerSemanticProperty("dfc-b:hasAddress", &method("address")).valueSetter = method("address=")
        registerSemanticProperty("dfc-b:hasGeoJsonFeature", &method("geo_json_feature")).valueSetter = method("geo_json_feature=")
        registerSemanticProperty("dfc-b:hasMainContact", &method("main_contact")).valueSetter = method("main_contact=")
        registerSemanticProperty("dfc-b:isOpenDuring", &method("is_open_during")).valueSetter = method("is_open_during=")
        registerSemanticProperty("dfc-b:localizes", &method("localizes")).valueSetter = method("localizes=")
        registerSemanticProperty("dfc-b:stores", &method("stores")).valueSetter = method("stores=")
      end

      Core::SemanticObject.type_registry[SEMANTIC_TYPE] = self
    end
  end
end

File: typescript-connector/src/models/PhysicalPlace.ts
// Class from DFC Business Ontology: #PhysicalPlace
import { SemanticObject } from "../core/SemanticObject.js";
import { Place, type PlaceParams } from "./Place.js";
import type { Address } from "./Address.js";
import type { Feature } from "./Feature.js";
import type { OpeningHoursSpecification } from "./OpeningHoursSpecification.js";
import type { Person } from "./Person.js";
import type { RealStock } from "./RealStock.js";
import type { TheoriticalStock } from "./TheoriticalStock.js";

export interface PhysicalPlaceParams extends PlaceParams {
  hasPhoneNumber?: string;
  hasAddress?: Address | string;
  hasGeoJsonFeature?: Feature | string;
  hasMainContact?: Person | string;
  isOpenDuring?: OpeningHoursSpecification | string;
  localizes?: (TheoriticalStock | string)[];
  stores?: (RealStock | string)[];
}
export class PhysicalPlace extends Place {
  static get SEMANTIC_TYPE(): string {
    return "dfc-b:PhysicalPlace";
  }

  hasPhoneNumber?: string;
  hasAddress?: Address | string;
  hasGeoJsonFeature?: Feature | string;
  hasMainContact?: Person | string;
  isOpenDuring?: OpeningHoursSpecification | string;
  localizes?: (TheoriticalStock | string)[];
  stores?: (RealStock | string)[];

  constructor(
    semanticId: string,
    params?: PhysicalPlaceParams,
  ) {
    super(semanticId, params);
    this.hasPhoneNumber = params?.hasPhoneNumber;
    this.hasAddress = params?.hasAddress;
    this.hasGeoJsonFeature = params?.hasGeoJsonFeature;
    this.hasMainContact = params?.hasMainContact;
    this.isOpenDuring = params?.isOpenDuring;
    this.localizes = params?.localizes;
    this.stores = params?.stores;
    this.semanticType = PhysicalPlace.SEMANTIC_TYPE;
    this.registerSemanticProperty("dfc-b:hasPhoneNumber", () => this.hasPhoneNumber);
    this.registerSemanticProperty("dfc-b:hasAddress", () => this.hasAddress);
    this.registerSemanticProperty("dfc-b:hasGeoJsonFeature", () => this.hasGeoJsonFeature);
    this.registerSemanticProperty("dfc-b:hasMainContact", () => this.hasMainContact);
    this.registerSemanticProperty("dfc-b:isOpenDuring", () => this.isOpenDuring);
    this.registerSemanticProperty("dfc-b:localizes", () => this.localizes);
    this.registerSemanticProperty("dfc-b:stores", () => this.stores);
  }
  static {
    SemanticObject.typeRegistry.set(PhysicalPlace.SEMANTIC_TYPE, PhysicalPlace);
  }
}
